### PR TITLE
Fix build warnings in src/Layers

### LINKED ISSUE #713: Compiler warnings elimination
Starting from the creation of this task, #116 is a subtask of this task.

MSVC:
- [ ] MSBuild
  - [ ] Enable "warnings are errors" mode *(is it possible?)*
- [ ] Compiler
  - [x] Enable conformance mode
  - [x] Enable "warnings are errors" mode
  - [ ] Raise the warning level to 4
    - [x] Level 1
    - [x] Level 2
    - [ ] Level 3
    - [ ] Level 4
- [ ] Linker
  - [ ] Enable "warnings are errors" mode

GCC:
- CMake and `make`
  - [ ] Enable "warnings are errors" mode *(is it possible?)*
- [ ] Compiler
  - [x] Disable permissive mode
  - [ ] *Maybe enable [pedantic mode](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html)*
  - [ ] Enable "warnings are errors" mode
  - [ ] Raise the warning level to 4
    - [ ] Level 1
    - [ ] Level 2
    - [ ] Level 3
    - [ ] Level 4
- [ ] Linker
  - [ ] Enable "warnings are errors" mode

For history:
- [ ] List all warnings elimination commits in this issue

Tools:
PVS-Studio
CppDepend (includes cppcheck, Clang-Tidy, Vera++)
ReSharper C++ (includes Clang and Clang-Tidy check plus it's own analysis)
Visual Assist (has it's own code inspection)

https://habr.com/ru/articles/75123/
https://habr.com/ru/companies/pvs-studio/articles/303664/

### PR DESCRIPTION
**Pull request by Masterkatze**
Part of #907

### CODE REVIEW

File: src/Layers/xrRenderPC_GL/rgl.h [hunk 1]
```diff
@@ -191,7 +191,7 @@ class CRender : public D3DXRenderBase
     R_sync_point q_sync_point;
 
     bool m_bMakeAsyncSS;
-    bool m_bFirstFrameAfterReset; // Determines weather the frame is the first after resetting device.
+    bool m_bFirstFrameAfterReset{ false }; // Determines weather the frame is the first after resetting device.
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/rgl.h]: When the default value is `0` or `false` or something like this, please, use `{}`.
```suggestion
    bool m_bFirstFrameAfterReset{}; // Determines weather the frame is the first after resetting device.
```

File: src/Layers/xrRenderPC_GL/r2_R_sun.cpp [hunk 2]
```diff
@@ -298,7 +298,8 @@ struct DumbClipper
             EFC_Visible res = frustum.testAABB(&bb.vMin.x, mask);
             switch (res)
             {
-            case fcvNone: continue;
+            case fcv_forcedword:
+                continue;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/r2_R_sun.cpp]: We agreed in Discord that we won't be adding all cases to switches.

More over, `fcv_forcedword` should be removed from that enum, just make `u32` underlying type.

File: src/Layers/xrRenderPC_GL/gl_rendertarget_u_set_rt.cpp [hunk 3]
```diff
@@ -27,8 +27,7 @@ void CRenderTarget::u_setrt(const ref_rt& _1, const ref_rt& _2, const ref_rt& _3
     }
     RCache.set_ZB(zb);
     //	RImplementation.rmNormal				();
-    GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
-    VERIFY(status == GL_FRAMEBUFFER_COMPLETE);
+    VERIFY(glCheckFramebufferStatus(GL_FRAMEBUFFER) == GL_FRAMEBUFFER_COMPLETE);
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/gl_rendertarget_u_set_rt.cpp]: Be cautious! VERIFY doesn't work on Release, while the logic of this code is that `glCheckFramebufferStatus` should be always executed, on Release too.
Correct change is this:
```suggestion
    [[maybe_unused]] GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
    VERIFY(status == GL_FRAMEBUFFER_COMPLETE);
```

File: src/Layers/xrRenderGL/glSH_RT.cpp [hunk 4]
```diff
@@ -22,18 +22,20 @@ CRT::~CRT()
 
 bool CRT::used_as_depth() const
 {
+    if (fmt == MAKEFOURCC('D', 'F', '2', '4'))
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderGL/glSH_RT.cpp]: Why it's moved to the top? I don't see any sense in this.

File: src/Layers/xrRenderGL/glBufferUtils.cpp [hunk 5]
```diff
@@ -212,7 +212,10 @@ void ConvertVertexDeclaration(u32 FVF, SDeclaration* decl)
 {
     RCache.set_Format(decl);
 
+#ifdef DEBUG
     GLsizei stride = GetFVFVertexSize(FVF);
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderGL/glBufferUtils.cpp]: Maybe, we can move GetFVFVertexSize(FVF) to the bottom, like this:
```cpp
VERIFY(offset == GetFVFVertexSize(FVF));

File: src/Layers/xrRender/dxStatGraphRender.cpp [hunk 6]
```diff
@@ -42,20 +42,18 @@ void dxStatGraphRender::OnRender(CStatGraph& owner)
...
+            break;
+        case CStatGraph::stCurve:
+            LineElem += it->elements.size() * 2;
+            break;
+        case CStatGraph::stBarLine:
+            LineElem += it->elements.size() * 4;
+            break;
+        //case CStatGraph::stPoint:
+        //    TriElem += it->elements.size() * 4;
+        //    break;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/dxStatGraphRender.cpp]:
```suggestion
        case CStatGraph::stPoint:
            // TriElem += it->elements.size() * 4;
            break;
```

File: src/Layers/xrRender/SkeletonAnimated.h [hunk 7]
```diff
@@ -75,9 +75,9 @@ class ECORE_API CKinematicsAnimated : public CKinematics, public IKinematicsAnim
 #else
 private:
 #endif
-    u32 Update_LastTime;
+    u32 Update_LastTime{ 0 };
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/SkeletonAnimated.h]:
```suggestion
    u32 Update_LastTime{};
```

File: src/Layers/xrRender/SkeletonAnimated.cpp [hunk 8]
```diff
@@ -632,9 +632,8 @@ void CKinematicsAnimated::Release()
 }
 
 CKinematicsAnimated::~CKinematicsAnimated() { IBoneInstances_Destroy(); }
-CKinematicsAnimated::CKinematicsAnimated()
-    : CKinematics(), IKinematicsAnimated(), blend_instances(nullptr), m_Partition(nullptr), m_blend_destroy_callback(nullptr),
-      m_update_tracks_callback(nullptr), Update_LastTime(0)
+
+CKinematicsAnimated::CKinematicsAnimated() : CKinematics(), IKinematicsAnimated()
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/SkeletonAnimated.cpp]: Looks like constructor is not needed at all.

File: src/Layers/xrRender_R2/r2_R_sun_support.h [hunk 9]
```diff
@@ -216,7 +216,7 @@ class FixedConvexVolume
             float min_dist = 2 * map_size;
             for (int p = 0; p < 4; ++p)
             {
-                float dist;
+                float dist = 0;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender_R2/r2_R_sun_support.h]: This is better:
```suggestion
                float dist = 0.f;
```
or
```suggestion
                float dist = 0.0f;
```

File: src/Layers/xrRenderPC_GL/gl_rendertarget_draw_rain.cpp [hunk 10]
```diff
@@ -147,12 +147,11 @@ void CRenderTarget::draw_rain(light& RainSetup)
         // clouds xform
         Fmatrix m_clouds_shadow;
         {
-            static float w_shift = 0;
             Fmatrix m_xform;
             //Fvector			direction	= RainSetup.direction	;
             Fvector normal;
             normal.setHP(1, 0);
-            //w_shift		+=	0.003f*Device.fTimeDelta;
+            //static float w_shift = 0.003f*Device.fTimeDelta;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/gl_rendertarget_draw_rain.cpp]: You changed the original code logic, please, leave the variable in the original place. Commented, though)

File: src/Layers/xrRenderPC_GL/r2_R_sun.cpp [hunk 11]
```diff
@@ -325,7 +326,7 @@ struct DumbClipper
                     }
                 }
                 break;
-            } // switch (res)
+            }
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/r2_R_sun.cpp]: The switch is big enough, it's better to leave the comment.
```suggestion
            } // switch (res)
```

File: src/Layers/xrRender_R2/r2_R_sun_support.h [hunk 12]
```diff
@@ -253,7 +253,7 @@ class FixedConvexVolume
                 continue;
             }
 
-            if (tmp_dist < 0 && orient < 0 || tmp_dist > 0 && orient > 0)
+            if ((tmp_dist < 0 && orient < 0) || (tmp_dist > 0 && orient > 0))
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender_R2/r2_R_sun_support.h]: Do you remember what we decided about parentheses overall?

File: src/Layers/xrRenderPC_GL/r3_R_rain.cpp [hunk 13]
```diff
@@ -232,8 +232,7 @@ void CRender::render_rain()
 
     // Begin SMAP-render
     {
-        bool bSpecialFull = mapNormalPasses[1][0].size() || mapMatrixPasses[1][0].size() || mapSorted.size();
-        VERIFY(!bSpecialFull);
+        VERIFY(!(mapNormalPasses[1][0].size() || mapMatrixPasses[1][0].size() || mapSorted.size()));
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/r3_R_rain.cpp]:
```suggestion
        VERIFY2(mapNormalPasses[1][0].empty() && mapMatrixPasses[1][0].empty() && mapSorted.empty(), "Special should be empty at this stage, but it's not empty...");
```

**Review by Xottab-DUTY — CHANGES_REQUESTED** [review]
Please, remove mimalloc and sse2neon changes from this PR

File: src/Layers/xrRender/BufferUtils.h [hunk 14]
```diff
@@ -78,17 +77,16 @@ class VertexStagingBuffer
 private:
     void Destroy();
 
-    VertexBufferHandle m_DeviceBuffer;
-    HostBufferHandle m_HostBuffer;
-    size_t m_Size{ 0 };
-    u32 m_RefCounter{ 0 };
-    bool m_AllowReadBack{ false }; // specifies whether host will want to have the data back (e.g. skinning code)
+	VertexBufferHandle m_DeviceBuffer{};
+	HostBufferHandle m_HostBuffer{};
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/BufferUtils.h]: You have tabs here...
image

File: src/Layers/xrRenderGL/glSH_RT.cpp [hunk 15]
```diff
@@ -29,11 +29,13 @@ bool CRT::used_as_depth() const
...
     case D3DFMT_D24X8:
     case D3DFMT_D24S8:
-    case MAKEFOURCC('D', 'F', '2', '4'):
-        return true;
-    default:
-        return false;
+        return true; 
     }
+
+	if (fmt == MAKEFOURCC('D', 'F', '2', '4'))
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderGL/glSH_RT.cpp]: MAKEFOURCC inside the switch cause warnings?(

File: src/Layers/xrRenderPC_GL/gl_rendertarget_draw_rain.cpp [hunk 16]
```diff
@@ -147,12 +147,12 @@ void CRenderTarget::draw_rain(light& RainSetup)
...
         Fmatrix m_clouds_shadow;
         {
-            static float w_shift = 0;
+            //static float w_shift = 0.0f;
             Fmatrix m_xform;
             //Fvector			direction	= RainSetup.direction	;
             Fvector normal;
             normal.setHP(1, 0);
-            //w_shift		+=	0.003f*Device.fTimeDelta;
+            //w_shift += 0.003f*Device.fTimeDelta;
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/gl_rendertarget_draw_rain.cpp]: This now doesn't correspond to the code style of old commented code :D
```suggestion
            //w_shift		+=	0.003f*Device.fTimeDelta;
```

File: src/Layers/xrRender/BufferUtils.h (hunk 14 above)
Comment (Masterkatze) [reply src/Layers/xrRender/BufferUtils.h]: oops

File: src/Layers/xrRenderGL/glSH_RT.cpp (hunk 15 above)
Comment (Masterkatze) [reply src/Layers/xrRenderGL/glSH_RT.cpp]: Yes, because there is no such item in _D3DFORMAT enum.

File: src/Layers/xrRender/Light_DB.cpp [hunk 17]
```diff
@@ -23,7 +23,7 @@ void CLight_DB::Load(IReader* fs)
         for (size_t i = 0; i < count; ++i)
         {
             Flight Ldata;
-            u32 controller = F->r_u32();
+            F->advance(sizeof(u32)); //controller
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/Light_DB.cpp]:
```suggestion
            F->advance(sizeof(u32)); // u32 controller = F->r_u32();
```

File: src/Layers/xrRenderGL/glSH_RT.cpp [hunk 18]
```diff
@@ -107,8 +109,8 @@ void CRT::resolve_into(CRT& destination) const
     RCache.set_RT(pRT, 0);
     RCache.set_RT(destination.pRT, 1);
 
-    const GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
-    VERIFY(status == GL_FRAMEBUFFER_COMPLETE);
+	[[maybe_unused]] GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
+	VERIFY(status == GL_FRAMEBUFFER_COMPLETE);
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderGL/glSH_RT.cpp]: Tabs -> spaces
```suggestion
    [[maybe_unused]] GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
    VERIFY(status == GL_FRAMEBUFFER_COMPLETE);
```

File: src/Layers/xrRenderGL/glSH_RT.cpp (hunk 15 above)
Comment (Xottab-DUTY) [inline src/Layers/xrRenderGL/glSH_RT.cpp]: Tab -> spaces
```suggestion
    if (fmt == MAKEFOURCC('D', 'F', '2', '4'))
```

File: src/Layers/xrRenderGL/glBufferUtils.cpp [hunk 19]
```diff
@@ -272,7 +271,7 @@ void ConvertVertexDeclaration(u32 FVF, SDeclaration* decl)
         offset += size * sizeof(float);
     }
 
-    VERIFY(stride == offset);
+	VERIFY(offset == GetFVFVertexSize(FVF));
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderGL/glBufferUtils.cpp]: Tab -> spaces
```suggestion
    VERIFY(offset == GetFVFVertexSize(FVF));
```

File: src/Layers/xrRenderGL/glBufferUtils.cpp [hunk 20]
```diff
@@ -212,7 +212,6 @@ void ConvertVertexDeclaration(u32 FVF, SDeclaration* decl)
 {
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderGL/glBufferUtils.cpp:212]: Maybe we should make FVF const here..

File: src/Layers/xrRenderPC_GL/gl_rendertarget_u_set_rt.cpp [hunk 21]
```diff
@@ -27,8 +27,8 @@ void CRenderTarget::u_setrt(const ref_rt& _1, const ref_rt& _2, const ref_rt& _3
     }
     RCache.set_ZB(zb);
     //	RImplementation.rmNormal				();
-    GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
-    VERIFY(status == GL_FRAMEBUFFER_COMPLETE);
+	[[maybe_unused]] GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/gl_rendertarget_u_set_rt.cpp]: Tabs -> spaces in this entire file (entire diff for this file)

File: src/Layers/xrRenderPC_GL/r2_R_sun.cpp [hunk 22]
```diff
@@ -869,9 +868,9 @@ void CRender::render_sun_near()
         FPU::m64r();
         // Lets begin from base frustum
 #ifdef	_DEBUG
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/r2_R_sun.cpp]: Tab to space :)
```suggestion
#ifdef _DEBUG
```

File: src/Layers/xrRenderGL/glSH_RT.cpp (hunk 15 above)
Comment (Xottab-DUTY) [reply src/Layers/xrRenderGL/glSH_RT.cpp]: Okay...
okay

File: src/Layers/xrRender_R2/r2.h [hunk 23]
```diff
@@ -192,8 +192,7 @@ class CRender final : public D3DXRenderBase
     R_sync_point q_sync_point;
 
     bool m_bMakeAsyncSS;
-    bool m_bFirstFrameAfterReset; // Determines weather the frame is the first after resetting device.
-
+    bool m_bFirstFrameAfterReset{}; // Determines weather the frame is the first after resetting device.
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender_R2/r2.h:195]:
```suggestion
    bool m_bFirstFrameAfterReset{}; // Determines weather the frame is the first after resetting device.

```

File: src/Layers/xrRenderDX10/3DFluid/dx103DFluidObstacles.cpp [hunk 24]
```diff
@@ -219,7 +219,7 @@ void dx103DFluidObstacles::ProcessDynamicObstacles(
         }
     }
 
-    if (!(m_lstShells.size() || m_lstElements.size()))
+    if (m_lstShells.empty() || m_lstElements.empty())
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderDX10/3DFluid/dx103DFluidObstacles.cpp]:
```suggestion
    if (m_lstShells.empty() && m_lstElements.empty())
```

**Review by Xottab-DUTY — APPROVED** [review]

**Comment by Xottab-DUTY** [comment]
Yay 🥳

**Pull request merged**